src/pipe_block.rs

```diff
@@ -0,0 +1,131 @@
+//! TODO: Header
+
+use std::sync::mpsc::{Receiver, Sender};
+use log::{error, info};
+
+/// A stage in the CyberGrape pipeline, which performs a step of the
+/// data aggregation, binauralization, or music playback process.
+pub trait Component {
+    type InData;
+    type OutData;
+    /// Converts an input of type A into an output of type B
+    fn convert(self: &Self, input: Self::InData) -> Self::OutData;
+}
+
+/// Runs on its own thread. On receiving data of type `I` on input, the
+/// PipeBlock converts them to data of type `O`, and sends it to its output
+/// channel.
+pub struct PipeBlock<'a, I, O> {
+    // component is an object that implments trait Component, and Rust cannot
+    // know how big component is at runtime. any 'dynamically typed' trait
+    // object must be marked with the `dyn` keyboard, post 2021
+
+    // the Box ensures that this PipeBlock owns the heap-alloc'd ref to the
+    // underlying Component
+
+    // the lifetime 'a indicates that the inner Component cannot live longer
+    // than the PipeBlock!
+    component: Box<dyn Component<InData = I, OutData = O> + 'a>,
+    output: Sender<O>,
+    input: Receiver<I>,
+}
+
+impl<'a, I, O> PipeBlock<'a, I, O> {
+    /// TODO
```

Comment:
Would be good to note here how we are supposed to create a `Component` to pass in